remkus/genesis-translations
# Switch tooling to npm and WP-CLI, automate creation of JavaScript translation files

## Summary
This pull request updates tooling to use WP-CLI and npm scripts.

It also automates creation of the new JavaScript translation files that Genesis 3.1 requires for WordPress 5.0+ to fully translate JavaScript strings.

- Removes Grunt in favour of npm and WP-CLI for automation (allows us to automate more and use WP-CLI language tooling without having to maintain and write our own Grunt plugins that do the same).
- Creates the plugin language file with WP-CLI instead of using Grunt (`wp i18n make-pot`).
- Clones Genesis from GitHub. (This is required temporarily to generate JavaScript language files with correct filenames — the `wp i18n make-json` command uses an md5 hash of the JavaScript file contents to generate the script language filename, so it needs to be able to read the Genesis source files.)
- Checks out the Genesis tag given in `genesisTranslations.genesisBranchOrTag` in `package.json` (currently 3.1.0).
- Uses a custom script to download Genesis translations from the StudioPress language server into `genesis-temp/lib/languages`. (I ported this script from `grunt-glotpress`, so the settings should seem familiar under the `glotpress` key in `package.json`.)
- Creates the JSON language files from the downloaded Genesis translations (`wp i18n make-json`).
- Copies all language files from the Genesis `lib/languages` directory to the plugin's `translations` directory.
- Renames JSON language files so that WordPress can find them. (WordPress expects the JSON language files to be prefixed by `genesis-`, but WP-CLI doesn't include this prefix.)
- Deletes the `genesis-temp` folder where Genesis was cloned to tidy things up.

## Requirements

1. Composer: https://getcomposer.org/doc/00-intro.md
2. Node.js: https://nodejs.org/en/download/current/
3. Access to the Genesis GitHub repo as your current git user.

## To test
1. Type `composer install` to install the WP-CLI i18n dependency.
2. Type `npm ci` to install JavaScript dependencies.
3. Set the Genesis tag you would like to use in the `genesisBranchOrTag` value in the package.json file. (Currently 3.1.0. Needs to be a valid branch of the Genesis repo, or one of these tags: https://github.com/studiopress/genesis/tags.)
4. Set the translation server version you would like to use in the glotpress.slug value in the package.json file. (Currently "genesis/genesis-3.1".)
5. Type `npm start`.

When the process completes you should see:

- [ ] Language files in the plugin's `translations` folder, including a large number of JSON files.
- [ ] Translations when activating the plugin and selecting a language in WordPress. This includes translations for the Genesis editor sidebar in Genesis 3.1 for supported languages (e.g. Spanish).

## Notes
- I made it so the download process completely removes files from the `translations` folder before adding new ones. This is to ensure that old files from previous Genesis versions do not remain. If this behaviour is not desired I can adjust it.
- @remkus You may need to adjust your `deploy.sh` script to exclude the new 'bin' and 'vendor' directories ('vendor' is created by composer).
- I have not included the Genesis 3.1 translations in this PR. A separate PR for this is here: https://github.com/remkus/genesis-translations/pull/11.
